From a3073db691e3c8c90394c3e2f8931e1570790c6e Mon Sep 17 00:00:00 2001
From: Dylan Hung <dylan_hung@aspeedtech.com>
Date: Fri, 6 Oct 2023 16:19:53 +0800
Subject: [PATCH 04/11] i3c: aspeed: Reorganize the APIs for entering and
 exiting the HALT state

In certain situations, such as receiving an error response or
experiencing a timeout during transfer transmission, it becomes
necessary to reset the hardware queues. This reset operation must be
executed when the controller is in the HALT state.

To achieve this, this commit adds the `enter_halt` API to handle these
two cases:
- Receiving an error response: the I3C controller will automatically
enter the HALT state. The API will monitore `present_state` register for
this.

- Timeout during transfer transmission: Use the `abort` bit to force the
controller to enter the HALT state.

When the reset is done, use `exit_halt` API to exit from the HALT state
and back to the normal operation state.

Signed-off-by: Dylan Hung <dylan_hung@aspeedtech.com>
Change-Id: Icb63e5cfd5732c2e2b16a1752455673f577cac46
---
 drivers/i3c/i3c_aspeed.c | 241 ++++++++++++++++++++++++++++++++++-----
 1 file changed, 214 insertions(+), 27 deletions(-)

diff --git a/drivers/i3c/i3c_aspeed.c b/drivers/i3c/i3c_aspeed.c
index 3c7d9964c8..6db8432728 100644
--- a/drivers/i3c/i3c_aspeed.c
+++ b/drivers/i3c/i3c_aspeed.c
@@ -275,7 +275,8 @@ union i3c_present_state_s {
 		volatile uint32_t current_master : 1;		/* bit[2] */
 		volatile uint32_t reserved0 : 5;		/* bit[7:3] */
 #define CM_TFR_STS_SLAVE_HALT	0x6
-#define CM_TFR_STS_MASTER_SERV_IBI	0xe
+#define CM_TFR_STS_MASTER_SERV_IBI 0xe
+#define CM_TFR_STS_MASTER_HALT	0xf
 		volatile uint32_t cm_tfr_sts : 6;		/* bit[13:8] */
 		volatile uint32_t reserved1 : 2;		/* bit[15:14] */
 		volatile uint32_t cm_tfr_st_sts : 6;		/* bit[21:16] */
@@ -524,6 +525,13 @@ struct i3c_aspeed_dev_priv {
 	} ibi;
 };
 
+enum i3c_aspeed_reset_type {
+	I3C_ASPEED_RESET_ALL,
+	I3C_ASPEED_RESET_QUEUES,
+	I3C_ASPEED_RESET_XFER_QUEUES,
+	I3C_ASPEED_RESET_IBI_QUEUE,
+};
+
 struct i3c_aspeed_obj {
 	const struct device *dev;
 	struct i3c_aspeed_config *config;
@@ -742,6 +750,106 @@ static void i3c_aspeed_rd_rx_fifo(struct i3c_aspeed_obj *obj, uint8_t *bytes, in
 	}
 }
 
+static int i3c_aspeed_reset_ctrl(struct i3c_aspeed_obj *obj, enum i3c_aspeed_reset_type type)
+{
+	struct i3c_aspeed_config *config = obj->config;
+	struct i3c_register_s *i3c_register = config->base;
+	union i3c_reset_ctrl_s reset_ctrl;
+	int ret;
+
+	reset_ctrl.value = 0;
+
+	switch (type) {
+	case I3C_ASPEED_RESET_IBI_QUEUE:
+		reset_ctrl.fields.ibi_queue_reset = 1;
+		break;
+	case I3C_ASPEED_RESET_ALL:
+		reset_ctrl.fields.core_reset = 1;
+		__fallthrough;
+	case I3C_ASPEED_RESET_QUEUES:
+		reset_ctrl.fields.ibi_queue_reset = 1;
+		__fallthrough;
+	case I3C_ASPEED_RESET_XFER_QUEUES:
+		reset_ctrl.fields.resp_queue_reset = 1;
+		reset_ctrl.fields.cmd_queue_reset = 1;
+		reset_ctrl.fields.tx_queue_reset = 1;
+		reset_ctrl.fields.rx_queue_reset = 1;
+		break;
+	default:
+		return -EINVAL;
+	}
+
+	i3c_register->reset_ctrl.value = reset_ctrl.value;
+	ret = reg_read_poll_timeout(i3c_register, reset_ctrl, reset_ctrl, !reset_ctrl.value, 0, 10);
+	if (ret) {
+		LOG_ERR("Reset failed. Reset ctrl: %08x %08x", reset_ctrl.value,
+			i3c_register->reset_ctrl.value);
+	}
+
+	return ret;
+}
+
+static void i3c_aspeed_exit_halt(struct i3c_aspeed_obj *obj)
+{
+	struct i3c_aspeed_config *config = obj->config;
+	struct i3c_register_s *i3c_register = config->base;
+	union i3c_device_ctrl_s ctrl;
+	union i3c_present_state_s state;
+	uint32_t halt_state = CM_TFR_STS_MASTER_HALT;
+	int ret;
+
+	if (obj->config->secondary) {
+		halt_state = CM_TFR_STS_SLAVE_HALT;
+	}
+
+	state.value = i3c_register->present_state.value;
+	if (state.fields.cm_tfr_sts != halt_state) {
+		LOG_DBG("I3C not in halt state, no need for resume");
+		return;
+	}
+
+	ctrl.value = i3c_register->device_ctrl.value;
+	ctrl.fields.resume = 1;
+	i3c_register->device_ctrl.value = ctrl.value;
+
+	ret = reg_read_poll_timeout(i3c_register, present_state, state,
+				    state.fields.cm_tfr_sts != halt_state, 0, 10);
+
+	if (ret) {
+		LOG_ERR("Exit halt state failed: %d %08x %08x", ret, state.value,
+			i3c_register->queue_status_level.value);
+	}
+}
+
+static void i3c_aspeed_enter_halt(struct i3c_aspeed_obj *obj, bool by_sw)
+{
+	struct i3c_aspeed_config *config = obj->config;
+	struct i3c_register_s *i3c_register = config->base;
+	union i3c_device_ctrl_s ctrl;
+	union i3c_present_state_s state;
+	uint32_t halt_state = CM_TFR_STS_MASTER_HALT;
+	int ret;
+
+	if (obj->config->secondary) {
+		halt_state = CM_TFR_STS_SLAVE_HALT;
+	}
+
+	LOG_DBG("present state = %08x\n", i3c_register->present_state.value);
+
+	if (by_sw) {
+		ctrl.value = i3c_register->device_ctrl.value;
+		ctrl.fields.abort = 1;
+		i3c_register->device_ctrl.value = ctrl.value;
+	}
+
+	ret = reg_read_poll_timeout(i3c_register, present_state, state,
+				    state.fields.cm_tfr_sts == halt_state, 0, 1000);
+	if (ret) {
+		LOG_ERR("Enter halt state failed: %d %08x %08x", ret, state.value,
+			i3c_register->queue_status_level.value);
+	}
+}
+
 static void i3c_aspeed_end_xfer(struct i3c_aspeed_obj *obj)
 {
 	struct i3c_register_s *i3c_register = obj->config->base;
@@ -774,15 +882,9 @@ static void i3c_aspeed_end_xfer(struct i3c_aspeed_obj *obj)
 	}
 
 	if (ret) {
-		union i3c_reset_ctrl_s reset_ctrl;
-
-		reset_ctrl.value = 0;
-		reset_ctrl.fields.rx_queue_reset = 1;
-		reset_ctrl.fields.tx_queue_reset = 1;
-		reset_ctrl.fields.resp_queue_reset = 1;
-		reset_ctrl.fields.cmd_queue_reset = 1;
-		i3c_register->reset_ctrl.value = reset_ctrl.value;
-		i3c_register->device_ctrl.fields.resume = 1;
+		i3c_aspeed_enter_halt(obj, false);
+		i3c_aspeed_reset_ctrl(obj, I3C_ASPEED_RESET_XFER_QUEUES);
+		i3c_aspeed_exit_halt(obj);
 	}
 
 	xfer->ret = ret;
@@ -980,7 +1082,8 @@ static void i3c_aspeed_slave_event(const struct device *dev, union i3c_intr_s st
 	if (status.fields.ccc_update) {
 		if (cm_tfr_sts == CM_TFR_STS_SLAVE_HALT) {
 			LOG_DBG("slave halt resume\n");
-			i3c_register->device_ctrl.fields.resume = 1;
+			i3c_aspeed_enter_halt(obj, false);
+			i3c_aspeed_exit_halt(obj);
 		}
 
 		if (i3c_register->slave_event_ctrl.fields.mrl_update) {
@@ -1006,6 +1109,7 @@ static void i3c_aspeed_isr(const struct device *dev)
 	struct i3c_aspeed_obj *obj = DEV_DATA(dev);
 	struct i3c_register_s *i3c_register = config->base;
 	union i3c_intr_s status;
+	uint32_t cm_tfr_sts;
 
 	status.value = i3c_register->intr_status.value;
 	if (config->secondary) {
@@ -1027,8 +1131,10 @@ static void i3c_aspeed_isr(const struct device *dev)
 		}
 	}
 
-	if (status.fields.xfr_error) {
-		i3c_register->device_ctrl.fields.resume = 1;
+	cm_tfr_sts = i3c_register->present_state.fields.cm_tfr_sts;
+	if (cm_tfr_sts == CM_TFR_STS_MASTER_HALT || cm_tfr_sts == CM_TFR_STS_SLAVE_HALT) {
+		LOG_ERR("Un-handled HALT");
+		i3c_aspeed_exit_halt(obj);
 	}
 
 	i3c_register->intr_status.value = status.value;
@@ -1404,7 +1510,12 @@ int i3c_aspeed_master_priv_xfer(struct i3c_dev_desc *i3cdev, struct i3c_priv_xfe
 	i3c_aspeed_start_xfer(obj, &xfer);
 
 	/* wait done, xfer.ret will be changed in ISR */
-	k_sem_take(&xfer.sem, I3C_ASPEED_XFER_TIMEOUT);
+	ret = k_sem_take(&xfer.sem, I3C_ASPEED_XFER_TIMEOUT);
+	if (ret) {
+		i3c_aspeed_enter_halt(obj, true);
+		i3c_aspeed_reset_ctrl(obj, I3C_ASPEED_RESET_XFER_QUEUES);
+		i3c_aspeed_exit_halt(obj);
+	}
 
 	/* report actual read length */
 	for (i = 0; i < nxfers; i++) {
@@ -1880,7 +1991,12 @@ int i3c_aspeed_master_send_ccc(const struct device *dev, struct i3c_ccc_cmd *ccc
 	i3c_aspeed_start_xfer(obj, &xfer);
 
 	/* wait done, xfer.ret will be changed in ISR */
-	k_sem_take(&xfer.sem, I3C_ASPEED_CCC_TIMEOUT);
+	ret = k_sem_take(&xfer.sem, I3C_ASPEED_CCC_TIMEOUT);
+	if (ret) {
+		i3c_aspeed_enter_halt(obj, true);
+		i3c_aspeed_reset_ctrl(obj, I3C_ASPEED_RESET_XFER_QUEUES);
+		i3c_aspeed_exit_halt(obj);
+	}
 
 	ret = xfer.ret;
 
@@ -1899,13 +2015,94 @@ static void sir_allowed_worker(struct k_work *work)
 	obj->sir_allowed_by_sw = 1;
 }
 
+static uint16_t parse_extra_gpio(const struct extra_gpio *extra_gpios, int size)
+{
+	const struct device *gpio_dev;
+	int i, ret;
+	uint16_t result = 0;
+
+	for (i = 0; i < size; i++) {
+		gpio_dev = device_get_binding(extra_gpios[i].port);
+		ret = gpio_pin_configure(gpio_dev, extra_gpios[i].pin, extra_gpios[i].flags | GPIO_INPUT);
+		if (ret < 0) {
+			LOG_ERR("pin %s:%d:%d configure failed %d\n", extra_gpios[i].port,
+				extra_gpios[i].pin, extra_gpios[i].flags | GPIO_INPUT, ret);
+			result = 0;
+			break;
+		}
+		ret = gpio_pin_get(gpio_dev, extra_gpios[i].pin);
+		if (ret < 0) {
+			LOG_ERR("pin %s:%d get value failed %d\n", extra_gpios[i].port,
+				extra_gpios[i].pin, ret);
+			result = 0;
+			break;
+		}
+		result |= ret << i;
+	}
+	LOG_DBG("extra val = %x\n", result);
+	return result;
+}
+
+int i3c_aspeed_master_send_entdaa(struct i3c_dev_desc *i3cdev)
+{
+	struct i3c_aspeed_obj *obj = DEV_DATA(i3cdev->bus);
+	struct i3c_aspeed_dev_priv *priv = DESC_PRIV(i3cdev);
+	struct i3c_aspeed_xfer xfer;
+	struct i3c_aspeed_cmd cmd;
+	union i3c_device_cmd_queue_port_s cmd_hi, cmd_lo;
+	int ret;
+
+	cmd_hi.value = 0;
+	cmd_hi.xfer_arg.cmd_attr = COMMAND_PORT_XFER_ARG;
+
+	cmd_lo.value = 0;
+	cmd_lo.addr_assign_cmd.cmd = I3C_CCC_ENTDAA;
+	cmd_lo.addr_assign_cmd.cmd_attr = COMMAND_PORT_ADDR_ASSIGN;
+	cmd_lo.addr_assign_cmd.toc = 1;
+	cmd_lo.addr_assign_cmd.roc = 1;
+
+	cmd_lo.addr_assign_cmd.dev_cnt = 1;
+	cmd_lo.addr_assign_cmd.dev_idx = priv->pos;
+
+	cmd.cmd_hi = cmd_hi.value;
+	cmd.cmd_lo = cmd_lo.value;
+	cmd.rx_length = 0;
+	cmd.tx_length = 0;
+
+	k_sem_init(&xfer.sem, 0, 1);
+	xfer.ret = -ETIMEDOUT;
+	xfer.ncmds = 1;
+	xfer.cmds = &cmd;
+	i3c_aspeed_start_xfer(obj, &xfer);
+
+	/* wait done, xfer.ret will be changed in ISR */
+	ret = k_sem_take(&xfer.sem, I3C_ASPEED_CCC_TIMEOUT);
+	if (ret) {
+		i3c_aspeed_enter_halt(obj, true);
+		i3c_aspeed_reset_ctrl(obj, I3C_ASPEED_RESET_XFER_QUEUES);
+		i3c_aspeed_exit_halt(obj);
+		return -ETIMEDOUT;
+	}
+
+	if (cmd.rx_length) {
+		LOG_INF("ENTDAA: No more new device. DA 0x%02x at DAT[%d] is not assigned",
+			i3cdev->info.dynamic_addr, priv->pos);
+		return -1;
+	}
+
+	i3c_master_send_getpid(i3cdev->bus, i3cdev->info.dynamic_addr, &i3cdev->info.pid);
+	LOG_INF("ENTDAA: DA 0x%02x at DAT[%d] is assigned, PID 0x%llx",
+			i3cdev->info.dynamic_addr, priv->pos, i3cdev->info.pid);
+
+	return 0;
+}
+
 static int i3c_aspeed_init(const struct device *dev)
 {
 	struct i3c_aspeed_config *config = DEV_CFG(dev);
 	struct i3c_aspeed_obj *obj = DEV_DATA(dev);
 	struct i3c_register_s *i3c_register = config->base;
 	const struct device *reset_dev = device_get_binding(ASPEED_RST_CTRL_NAME);
-	union i3c_reset_ctrl_s reset_ctrl;
 	union i3c_intr_s intr_reg;
 	int ret;
 
@@ -1916,17 +2113,7 @@ static int i3c_aspeed_init(const struct device *dev)
 	reset_control_deassert(reset_dev, config->reset_id);
 
 	ret = i3c_aspeed_disable(obj);
-
-	reset_ctrl.value = 0;
-	reset_ctrl.fields.core_reset = 1;
-	reset_ctrl.fields.tx_queue_reset = 1;
-	reset_ctrl.fields.rx_queue_reset = 1;
-	reset_ctrl.fields.ibi_queue_reset = 1;
-	reset_ctrl.fields.cmd_queue_reset = 1;
-	reset_ctrl.fields.resp_queue_reset = 1;
-	i3c_register->reset_ctrl.value = reset_ctrl.value;
-
-	ret = reg_read_poll_timeout(i3c_register, reset_ctrl, reset_ctrl, !reset_ctrl.value, 0, 10);
+	ret = i3c_aspeed_reset_ctrl(obj, I3C_ASPEED_RESET_ALL);
 	if (ret) {
 		return ret;
 	}
-- 
2.25.1

